File: examples/python/temporal_pipeline.py
"""temporal stride + seek example.

video understanding models (SlowFast, TimeSformer, Video Swin, X3D) need
frames sampled across time, not consecutive frames. temporal_stride(N)
collects every Nth decoded frame into each batch.
"""
import cuframe

pipeline = (cuframe.Pipeline.builder()
    .input("video.mp4")
    .resize(224, 224)
    .normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    .batch(16)
    .temporal_stride(4)  # every 4th frame
    .build())

print(f"source: {pipeline.source_width}x{pipeline.source_height}, "
      f"{pipeline.fps:.1f} fps, {pipeline.frame_count} frames")

# iterate clips — each batch has 16 frames spaced 4 apart
for i, clip in enumerate(pipeline):
    # torch.from_dlpack(clip) → [count, 3, 224, 224] CUDA tensor
    print(f"clip {i}: {clip.count} frames")

# random access: sample specific timestamps
for t in [1.0, 5.0, 10.0]:
    pipeline.seek(t)
    batch = pipeline.next()
    if batch is not None:
        # torch.from_dlpack(batch) for inference
        print(f"seek({t}): got {batch.count} frames")
    else:
        print(f"seek({t}): past end of video")

File: examples/temporal_pipeline.cpp
#include <cuframe/cuframe.h>
#include <cstdio>

int main(int argc, char** argv) {
    if (argc < 2) {
        printf("usage: %s <video>\n", argv[0]);
        return 1;
    }

    // video understanding: 16 frames sampled every 4th frame
    auto pipeline = cuframe::Pipeline::builder()
        .input(argv[1])
        .resize(224, 224)
        .normalize({0.485f, 0.456f, 0.406f}, {0.229f, 0.224f, 0.225f})
        .batch(16)
        .temporal_stride(4)
        .build();

    printf("source: %dx%d, %.1f fps, %lld frames\n",
           pipeline.source_width(), pipeline.source_height(),
           pipeline.fps(), pipeline.frame_count());

    int clip_idx = 0;
    while (auto clip = pipeline.next()) {
        // (*clip)->data() is float* to [count, 3, 224, 224] NCHW tensor on GPU
        // each frame is 4 decoded frames apart in time
        printf("clip %d: %d frames (spanning %d decoded frames)\n",
               clip_idx++, (*clip)->count(),
               (*clip)->count() * 4);
    }
    printf("total clips: %d\n", clip_idx);

    // random access: seek back to 1 second and grab a clip
    pipeline.seek(1.0);
    if (auto clip = pipeline.next()) {
        printf("after seek(1.0): clip with %d frames\n", (*clip)->count());
    }
}
